llama-cpp.nix

```diff
@@ -0,0 +1,118 @@
+, rocmPackages
+, clblast
+, Accelerate ? null
+, MetalKit ? null
+, CoreVideo ? null
+, CoreGraphics ? null
+, useOpenCL ? false
+, useCuda ? false
+, useRocm ? false
+}:
+
+let
+  inherit (lib) optional optionals versionOlder;
+  isDefault = !useOpenCL && !useCuda && !useRocm;
+
+  # Give a little description difference between the flavors.
+  descriptionSuffix = if useOpenCL then
+    " (OpenCL accelerated)"
+  else if useCuda then
+    " (CUDA accelerated)"
+  else if useRocm then
+    " (ROCm accelerated)"
+  else if (MetalKit != null) then
+    " (MetalKit accelerated)"
+  else "";
+
+  llama-python = python3.withPackages (ps: [ ps.numpy ps.sentencepiece ]);
+
+  # See ./overlay.nix for where these dependencies are passed in.
+  defaultBuildInputs = builtins.filter (p: p != null) [
+    Accelerate
+    MetalKit
+    CoreVideo
+    CoreGraphics
+  ];
+
+  cudaBuildInput = symlinkJoin (with cudaPackages; {
+    # HACK(Green-Sky): Nix currently has issues with CMake findcudatoolkit
+    # see https://github.com/NixOS/nixpkgs/issues/224291
+    # copied from jaxlib
+    name = "${cudatoolkit.name}-merged";
+    paths = [
+      cudatoolkit.lib
+      cudatoolkit.out
+    ] ++ optionals (versionOlder cudatoolkit.version "11") [
+      # For some reason, some of the required libs are in the targets/x86_64-linux
+      # directory; not sure why, but this works around it.
+      "${cudatoolkit}/targets/${stdenv.system}"
+    ];
+  });
+
+  rocmBuildInputs = with rocmPackages; [ clr hipblas rocblas ];
+in
+
+stdenv.mkDerivation {
+  name = "llama.cpp";
+  src = ./.;
+  meta = {
+    description = "Inference of LLaMA model in pure C/C++${descriptionSuffix}";
+    mainProgram = "llama";
+  };
+
+  postPatch = ''
+    substituteInPlace ./ggml-metal.m \
+      --replace '[bundle pathForResource:@"ggml-metal" ofType:@"metal"];' "@\"$out/bin/ggml-metal.metal\";"
+    substituteInPlace ./*.py --replace "/usr/bin/env python" "${llama-python}/bin/python"
+  '';
+
+  nativeBuildInputs = [ cmake ninja pkg-config ];
+
+  buildInputs = [ openmpi ]
+  ++ optional useOpenCL clblast
+  ++ optional useCuda cudaBuildInput
+  ++ optionals useRocm rocmBuildInputs
+  ++ optionals isDefault defaultBuildInputs;
+
+  cmakeFlags = [
+    "-DLLAMA_NATIVE=OFF"
+    "-DLLAMA_BUILD_SERVER=ON"
+    "-DBUILD_SHARED_LIBS=ON"
+    "-DCMAKE_SKIP_BUILD_RPATH=ON"
+  ]
+  ++ optional useOpenCL "-DLLAMA_CLBLAST=ON"
+  ++ optional useCuda "-DLLAMA_CUBLAS=ON"
+  ++ optionals useRocm [
+    "-DLLAMA_HIPBLAS=1"
+    "-DCMAKE_C_COMPILER=hipcc"
+    "-DCMAKE_CXX_COMPILER=hipcc"
+    # Build all targets supported by rocBLAS. When updating search for TARGET_LIST_ROCM
+    # in https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/CMakeLists.txt
+    # and select the line that matches the current nixpkgs version of rocBLAS.
+    "-DAMDGPU_TARGETS=gfx803;gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a:xnack+;gfx90a:xnack-;gfx940;gfx941;gfx942;gfx1010;gfx1012;gfx1030;gfx1100;gfx1101;gfx1102"
```

Comment:
Nit: make it a list and an argument to `callPackage`. Looking at the expression for pytorch in Nixpkgs, you could take `rocmPackages.clr.gpuTargets` as the default value.

Reply:
On this note, if you want this package to respect `config.cudaCapabilities`, you currently have to specify those manually as well (we might automate this for FindCUDAToolkit.cmake later)

Reply:
```
$ nix eval nixpkgs#rocmPackages.clr.gpuTargets
[ "gfx803" "gfx900" "gfx906" "gfx908" "gfx90a" "gfx940" "gfx941" "gfx942" "gfx1010" "gfx1012" "gfx1030" "gfx1100" "gfx1101" "gfx1102" ]
```

It's missing the `xnack-` entries, but the other entries are the same. I think I'll leave this to someone who actually owns a Radeon card and knows what's going on with ROCm, though I agree that this looks super gross right now.

Reply:
I suggest just using ``rocmPackages.clr.gpuTargets`` and appending the xnack variants.